openedx/frontend-app-communications
# feat: slot component for dynamic plugins

### Description

Following this [POC pluggable approach](https://github.com/openedx/frontend-app-course-authoring/pull/638) to importing plugins, we've created a reusable component that will dynamically import a module wherever it is called. The main idea is to be able to change the plugins with this mechanism, which means that instead of modifying the micro-frontend beforehand, we could change the plugin in isolation.

![slot-compoentn](https://github.com/openedx/frontend-app-communications/assets/134975835/ad1cb179-320c-4e2f-a1c4-f9f2d00ac51c)

---
Using pluggable component with `BulkEmailForm` component

![image](https://github.com/openedx/frontend-app-communications/assets/134975835/b73db260-b524-49c9-8741-364fcdd82870)


---
The alert is a plugin as well
![image](https://github.com/openedx/frontend-app-communications/assets/134975835/6b2e483c-2846-45d9-83d8-3f51fa044f5d)



#### How does it work?

Any component can be wrapped with the pluggable component. For instance:

```jsx
<PluggableComponent id="input-form" as="communications-app-card">
    <h1>Any component as default</h1>
</PluggableComponent>
```

The prop `as` indicates where is the plugin allocated. In this case, the component will be pointing to this route: `node_modules/@openedx-plugins/communications-app-card`.

If the plugin was installed as a dependency, it will render the component that you have pointing to that one.

If the plugin isn't installed, it will return its children. In this case, it's this:
```jsx
<h1>Any component as default</h1>
```

### Multiple plugins

We've added these props to `PluggableComponent`:

- `plugins`: This prop allows us to add multiple plugins, turning `PluggableComponent` into a container where its children will be rendered. Here's an example:

```jsx
import PluggableComponent from 'path-to/PluggableComponent';

const myPluginsList = [
  { id: 'plugin1', name: 'communications-app-test-component' },
  { id: 'plugin2', name: 'communications-app-test-component-2' }
];

const MultiplePlugins = () => (
  <PluggableComponent plugins={myPluginsList} />
);
```

If the name of the plugin, for example, is @openedx-plugins/communications-app-test-component, you should provide the name without @openedx-plugins, so it would be "communications-app-test-component".

- `pluginsPrefix`:  This prop allows us to add multiple plugins by specifying a prefix. PluggableComponent will find all plugins with that prefix and render them. For example, suppose you have three plugins:
    - @openedx-plugins/example-app-plugins-first-one
    - @openedx-plugins/example-app-plugins-second-one
    - @openedx-plugins/example-app-plugins-third-one
    
You only need to specify the prefix as a prop of PluggableComponent. In this case, "example-app-plugins" will render all plugins installed with that prefix. Here's an example:

```jsx
 import PluggableComponent from 'path-to/PluggableComponent';

  const MultiplePlugins = () => (
    <PluggableComponent pluginsPrefix="example-app-plugins"  />
  );

```
- `containerPluginsProps`: This prop allows us to specify props for the container of the plugins, such as a className, styles, or other props. Here's an example:

```jsx
 import PluggableComponent from 'path-to/PluggableComponent';

  const MultiplePlugins = () => (
    <PluggableComponent 
       pluginsPrefix="example-app-plugins" 
       containerPluginsProps={ className: "my_class", style: { backgroundColor: "red" } }  
    />
  );

```


#### Why are we doing this?

This means that the plugin doesn't have to be installed if I'm making a feature. It's something that we can improve in the future. That way, another developer can replace that component and make more changes. This approach allows us to wrap as many components as we can, making most of the features pluggable and adaptable to various use cases. The component can be rendered without children as well, which means it won't return anything if the plugin is not installed.


### How to test it

Check it out this branch and follow these steps:

```
 nvm use && npm install
```
Then you can run the mfe with `npm start`

If you want to create another plugin you can do it in the folder plugins/communications-app/ with the structure of the other
plugins

### Adding a Plugin That Lives Outside This Repository
You can do it by pointing to your npm package in your ``package.json``

```json
{
  "dependencies": {
     "@openedx-plugins/communications-app-body-email-form": "npm:@openedx/plugins-communications-app-body-email-form@^1.0.0"
  }
}
```
Or installing it like this

```bash
npm install --legacy-peer-deps "@openedx-plugins/communications-app-body-email-form@npm:@openedx/plugins-communications-app-body-email-form@^1.0.0"
```

To create and upload a plugin to npm you could follow this [guide](https://github.com/openedx/frontend-app-communications/blob/fc0b1bf2d18146fd626424009e8e92d10770ce87/docs/how_tos/create_an_external_plugin_ui_slot.rst)
